docs/rfc/rfc-011-delete-gas.md

```diff
@@ -0,0 +1,140 @@
+with the highest listed gas fees into each block. This produces a naturally ordering
+of the pending transactions. Applications wishing to implement a gas mechanism need
+to be able to order the transactions in the mempool. This can trivially be accomplished 
+by sorting transactions using the `priority` field available to applications as part of
+v0.35's `ResponseCheckTx` message.
+
+### Requirement: Allow Application-Defined Block Resizing
+
+When creating a block proposal, Tendermint pulls a set of possible transactions out of
+the mempool to include in the next block. Tendermint uses MaxGas to limit the set of transactions
+it pulls out of the mempool fetching a set of transactions whose sum is less than MaxGas.
+
+By removing gas tracking from Tendermint's mempool, Tendermint will need to provide a way for
+applications to determine an acceptable set of transactions to include in the block.
+
+This is what the new ABCI++ `PrepareProposal` method is useful for. Applications
+that wish to limit the contents of a block by an application-defined limit may
+do so by removing transactions from the proposal it is passed during `PrepareProposal`.
+
+Fully meeting the requirements of the current Gas system would mean adding functionality that is
+not currently provided the `PrepareProposal` method.
+Applications that wish to include _more_ transactions in the proposal than `PrepareProposal`
+initially passed to it cannot do so in the current design. [Issue 7750][issue-7750] has been filed to discuss
+and possibly solve for this.
+
+### Requirement: Handle Transaction Metadata
+
+Moving the gas mechanism into applications adds an additional piece of complexity
+to applications. The application must now track how much gas it expects a transaction
+to consume. The mempool currently handles this bookkeeping responsibility and uses the estimated
+gas to determine the set of transactions to include in the block. In order to task
+the application with keeping track of this metadata, we should make it easier for the
+application to do so. In general, we'll want to keep only one copy of this type
+of metadata in the program at a time, either in the application or in Tendermint.
+
+The following sections are possible solutions to the problem of storing transaction
+metadata without duplication.
+
+#### Metadata Handling: EvictTx Callback
+
+A possible approach to handling transaction metadata is by adding a new `EvictTx`
+ABCI method. Whenever the mempool is removing a transaction, either because it has
+reached its TTL or because it failed `RecheckTx`, `EvictTx` would be called with
+the transaction hash. This would indicate to the application that it could free any
+metadata it was storing about the transaction such as the computed gas fee.
+
+Eviction callbacks are pretty common in caching systems, so this would be very
+well-worn territory.
+
+#### Metadata Handling: Application-Specific Metadata Field(s)
```

Comment:
This seems like a more reasonable approach to me: the responsibility for caching stays within Tendermint, and the application can optionally do some arbitrary (yet deterministic) precomputation on the transactions during `CheckTx` before they're sent to `PrepareProposal`.

The challenge with this approach is that an application could inflate the mempool size by arbitrary amounts, which, if there aren't sane safety limits built into Tendermint, could cause problems.

Reply:
I'd also add here that I think the node-side transaction queue should probably not have a "size limit": If we buffer some amount in memory for efficiency, great, but when we exceed that I think we should get away from discarding transactions that the application hasn't passed judgement on. If that means we have to (sometimes) spill data to a file somewhere, it would slow things down a little bit until the app catches up, but needn't OOM the process.

Of course, the application could still blow up RAM, and if it lives in-process with Tendermint that takes the node too, but as you pointed out, that's not something we can really control directly.

Reply:
> I'm not sure I agree with this yet and think it warrants continued discussion. This effectively means maintaining two unbounded queues: 1 queue of Tx that have not been seen by the app via `CheckTx` and a second queue that have.

You raise a very good point here: Mitigating bad behaviour (to the extent possible) is also an important concern.

To some extent we already have this problem now, only from the other side: If bad actors spam a lot of txn, they could potentially cause txn belonging to good-faith participants to be evicted or superseded. Defending against that seems like a concern we have either way.
 
Is a size limit the right way to protect against this? Or, perhaps, should the node be keeping track of other properties of the transaction, e.g., "I already have 250 distinct txn from this one peer, I think they're spammy and am going to drop theirs".

Among txn we think are in "good faith", though, I'd expect the app to at least participate in their disposition—otherwise infrastructural attacks can not just slow or halt consensus (which is bad), they could potentially affect the content of blocks (which seems to me like a much worse attack).

Reply:
> Is a size limit the right way to protect against this? Or, perhaps, should the node be keeping track of other properties of the transaction, e.g., "I already have 250 distinct txn from this one peer, I think they're spammy and am going to drop theirs".

Yeah, I would agree that this is a reasonably carveout to make. It does possibly add a case where a Tx can be dropped for a reason that the application did not know about. An option would also be to disconnect from peers that frequently send invalid (from the point of view of CheckTx) transactions. I think that we have a pretty good baseline for our understanding of the problem of Gas and think we can discuss that on further RFCs.